cpp/src/io/functions.cpp

```diff
@@ -106,76 +106,113 @@ chunked_parquet_writer_options_builder chunked_parquet_writer_options::builder(
 }
 
 namespace {
-template <typename reader, typename reader_options>
-std::unique_ptr<reader> make_reader(source_info const& src_info,
-                                    reader_options const& options,
-                                    rmm::cuda_stream_view stream,
-                                    rmm::mr::device_memory_resource* mr)
-{
-  if (src_info.type == io_type::FILEPATH) {
-    return std::make_unique<reader>(src_info.filepaths, options, stream, mr);
-  }
 
-  std::vector<std::unique_ptr<datasource>> datasources;
-  if (src_info.type == io_type::HOST_BUFFER) {
-    datasources = cudf::io::datasource::create(src_info.buffers);
-  } else if (src_info.type == io_type::USER_IMPLEMENTED) {
-    datasources = cudf::io::datasource::create(src_info.user_sources);
-  } else {
-    CUDF_FAIL("Unsupported source type");
+std::vector<std::unique_ptr<cudf::io::datasource>> make_datasources(source_info const& info,
```

Comment:
What was the motivation behind switching from readers/writers to sources/sinks? Is it that readers/writers imply a file system, and that actual data streams needn't be files?
If yes, might I please enquire what system might use, say, Parquet serialization for streams that aren't files?

Reply:
> what system might use, say, Parquet serialization for streams that aren't files?

Ah, that's the wrong question. The stream could be a CSV or JSON stream that didn't start out as a file.

Reply:
we didn't really "change" from readers/writers to source/sinks. We still create the appropriate reader/writer, just after this function is called. The previous function was forcing all readers or writers to have the same constructor, though, and that was not very helpful. This way the calling function has more control over how to create the reader/writer.